packages/flame/lib/src/text/sprite_font_renderer.dart

```diff
@@ -0,0 +1,163 @@
+/// write text using these sprite images as the font.
+///
+/// Currently, this class supports monospace fonts only -- the width and the
+/// height of all characters must be the same ([charWidth] and [charHeight]).
+/// Extra space between letters can be added via the [letterSpacing] parameter
+/// (it can also be negative to "squish" characters together). Finally, the
+/// [scale] parameter allows scaling the font to be bigger/smaller relative to
+/// its size in the source image.
+///
+/// The [paint] parameter is used to composite the character images onto the
+/// canvas. Its default value will draw the character images as-is. Changing
+/// the opacity of the paint's color will make the text semi-transparent.
+class SpriteFontRenderer extends TextRenderer {
+  SpriteFontRenderer._({
+    required this.source,
+    required this.charWidth,
+    required this.charHeight,
+    required Map<int, _GlyphInfo> glyphs,
+    required this.letterSpacing,
+    required this.scale,
+  }) : _glyphs = glyphs;
+
+  final Image source;
+  final double charWidth;
+  final double charHeight;
+  final double scale;
+  final double letterSpacing;
+  bool get isMonospace => true;
+
+  final Map<int, _GlyphInfo> _glyphs;
+  Paint paint = Paint()..color = const Color(0xFFFFFFFF);
+
+  @override
+  double measureTextHeight(String text) => charHeight * scale;
+
+  @override
+  double measureTextWidth(String text) {
+    final n = text.length;
+    return n > 0 ? charWidth * scale * n + letterSpacing * (n - 1) : 0;
+  }
+
+  @override
+  Vector2 measureText(String text) {
+    return Vector2(measureTextWidth(text), measureTextHeight(text));
+  }
+
+  @override
+  void render(
+    Canvas canvas,
+    String text,
+    Vector2 position, {
+    Anchor anchor = Anchor.topLeft,
+  }) {
+    final rstTransforms = Float32List(4 * text.length);
+    final rects = Float32List(4 * text.length);
+    var j = 0;
+    var x0 = position.x;
+    final y0 = position.y;
+    for (final glyph in _textToGlyphs(text)) {
+      rects[j + 0] = glyph.srcLeft;
+      rects[j + 1] = glyph.srcTop;
+      rects[j + 2] = glyph.srcRight;
+      rects[j + 3] = glyph.srcBottom;
+      rstTransforms[j + 0] = glyph.rstSCos;
+      rstTransforms[j + 1] = glyph.rstSSin;
+      rstTransforms[j + 2] = x0 + glyph.rstTx;
+      rstTransforms[j + 3] = y0 + glyph.rstTy;
+      x0 += glyph.width + letterSpacing;
+      j += 4;
+    }
+    canvas.drawRawAtlas(source, rstTransforms, rects, null, null, null, paint);
+  }
+
+  Iterable<_GlyphInfo> _textToGlyphs(String text) {
+    return text.codeUnits.map(_getGlyphFromCodeUnit);
+  }
+
+  _GlyphInfo _getGlyphFromCodeUnit(int i) {
+    final glyph = _glyphs[i];
+    if (glyph == null) {
+      throw ArgumentError('No glyph for character "${String.fromCharCode(i)}"');
+    }
+    return glyph;
+  }
+}
+
+/// Helper class for creating [SpriteFontRenderer] instances.
+class SpriteFontBuilder {
```

Comment:
I really think we should avoid the builder pattern, it is something not common on Flutter/Dart and on Flame we don't use that as well.

A solution more in line with, for example, how we made the `SpriteAnimation` constructor would be preferable. Where we have a metadata class that describe the animation and internally it is used to load the frames. Here we could have the same thing, a class for a glyph metadata and `SpriteFontRenderer` will receive a map of it.

That we keep the API more declarative and more in line with the APIs we current have.

Reply:
Alright, I changed it into `SpriteFontData`.

Reply:
Alright, I changed it into `SpriteFontData`.